// PinguPackage/src/pingu/user/items/Achievement.ts
import { PItem } from "../../../database/json/PItem";
import { PinguClientEvents } from "../../handlers/PinguEvent";

export type AchievementType = 'GUILD' | 'USER'
export type UserAchievementTypes = 'CHAT' | 'COMMAND'
export type GuildAchievementTypes = UserAchievementTypes | 'VOICE'

class AchievementBase extends PItem {
    constructor(id: number, name: string, type: AchievementType, subType: UserAchievementTypes, description: string, ) {
        super({ id: id.toString(), name });
        this.type = type;
        this.subType = subType;
        this.description = description;
    }

    public type: AchievementType
    public subType: UserAchievementTypes
    public description: string
}

export class UserAchievement extends AchievementBase {
    constructor(id: number, name: string, type: UserAchievementTypes, description: string) {
        super(id, name, 'USER', type, description);
    }
}
export class Achievement extends PItem {
    constructor(id: number, name: string, type: AchievementType, description: string, ) {
        super({ id: id.toString(), name });
    }

    public static Achievements: Achievement[] = [
        new Achievement(1, )
    ]
}
