// PinguPackage/live/database/json/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PClient = exports.PRole = exports.PQueue = exports.PMarry = exports.PUser = exports.PGuildMember = exports.PGuild = exports.PChannel = exports.PItem = void 0;
var PItem_1 = require("./PItem");
Object.defineProperty(exports, "PItem", { enumerable: true, get: function () { return PItem_1.PItem; } });
var PChannel_1 = require("./PChannel");
Object.defineProperty(exports, "PChannel", { enumerable: true, get: function () { return PChannel_1.PChannel; } });
var PGuild_1 = require("./PGuild");
Object.defineProperty(exports, "PGuild", { enumerable: true, get: function () { return PGuild_1.PGuild; } });
var PGuildMember_1 = require("./PGuildMember");
Object.defineProperty(exports, "PGuildMember", { enumerable: true, get: function () { return PGuildMember_1.PGuildMember; } });
var PUser_1 = require("./PUser");
Object.defineProperty(exports, "PUser", { enumerable: true, get: function () { return PUser_1.PUser; } });
var PMarry_1 = require("./PMarry");
Object.defineProperty(exports, "PMarry", { enumerable: true, get: function () { return PMarry_1.PMarry; } });
var PQueue_1 = require("./PQueue");
Object.defineProperty(exports, "PQueue", { enumerable: true, get: function () { return PQueue_1.PQueue; } });
var PRole_1 = require("./PRole");
Object.defineProperty(exports, "PRole", { enumerable: true, get: function () { return PRole_1.PRole; } });
var PClient_1 = require("./PClient");
Object.defineProperty(exports, "PClient", { enumerable: true, get: function () { return PClient_1.PClient; } });
