// PinguPackage/src/MongoSchemas/PinguGuild.ts
import * as mongoose from 'mongoose';
import {PinguGuild} from '../pingu/guild/PinguGuild';

const PItem = {
    _id: String,
    name: String
};
const DecidableConfig = {
    firstTimeExecuted: Boolean,
    channel: PItem
};
const DecidableItem = {
    value: String,
    _id: String,
    author: PItem,
    channel: PItem,
    endsAt: Date
};

const PinguGuildSchema = mongoose.model('PinguGuild', new mongoose.Schema({
    _id: String,
    name: String,
    guildOwner: PItem,
    clients: [{
        _id: String,
        embedColor: Number,
        prefix: String,
        displayName: String
    }],
    welcomeChannel: PItem,
    reactionRoles: [{
        channel: PItem,
        messageID: String,
        emoteName: String,
        pRole: PItem
    }],
    giveawayConfig: { ...DecidableConfig,
        allowSameWinner: Boolean,
        hostRole: PItem,
        winnerRole: PItem,
        giveaways: [{ ...DecidableItem,
            winners: [PItem]
        }],
    },
    pollConfig: { ...DecidableConfig,
        pollsRole: PItem,
        polls: [{ ...DecidableItem, 
            YesVotes: Number,
            NoVotes: Number,
            approved: String
        }]
    },
    suggestionConfig: { ...DecidableConfig, 
        verifyRole: PItem,
        suggestions: [{ ...DecidableItem,
            approved: String,
            decidedBy: PItem
        }]
        
    },
    themeWinners: [PItem]
})) as mongoose.Model<mongoose.Document<PinguGuild>>

export {PinguGuildSchema };
