// PinguPackage/src/pingu/library/logChannels.ts
import { Client, Message, MessageEmbed, User } from "discord.js";
import { Error } from "../../helpers";
import { PinguLibrary } from "./PinguLibrary";
import { PinguEvents } from "../PinguEvents";
import * as config from '../../config.json';
const { version } = config;


export async function errorLog(client: Client, message: string, messageContent?: string, err?: Error) {
    var errorlogChannel = PinguLibrary.getChannel(client, PinguLibrary.SavedServers.PinguSupport(client).id, 'error-log-⚠');
    if (!errorlogChannel) return PinguLibrary.DanhoDM(client, 'Unable to find #error-log-⚠️ in Pingu Support, https://discord.gg/gbxRV4Ekvh');

    console.error(getErrorMessage(message.includes('`') ? message.replace('`', ' ') : message, messageContent, err));
    let sent = await errorlogChannel.send(getErrorMessage(message, messageContent, err));
    await sent.react(PinguLibrary.SavedServers.DanhoMisc(client).emojis.cache.find(e => e.name == 'Checkmark'));
    return sent;

    function getErrorMessage(message: string, messageContent?: string, err?: Error) {
        let result = {
            format: "```\n",
            providedMessage: `[Provided Message]\n${message}\n\n`,
            errorMessage: `[Error message]: \n${err && err.message}\n\n`,
            messageContent: `[Message content]\n${messageContent}\n\n`,
            stack: `[Stack]\n${err && err.stack}\n\n\n`,
            fileMessage: `${err && err.fileName} threw an error at line ${err && err.lineNumber}!\n\n`
        };

        let returnMessage = (
            result.format +
            (err && err.fileName && err.lineNumber ? result.fileMessage : "") +
            result.providedMessage +
            (messageContent ? result.messageContent : "") +
            (err ? result.errorMessage + result.stack : "") +
            result.format
        );

        PinguLibrary.consoleLog(client, returnMessage);
        return returnMessage
    }
}
export async function pGuildLog(client: Client, script: string, message: string, err ?: Error) {
    var pinguGuildLog = PinguLibrary.getChannel(client, PinguLibrary.SavedServers.PinguSupport(client).id, "pingu-guild-log-🏡");
    if (!pinguGuildLog) return PinguLibrary.DanhoDM(client, `Couldn't get #pingu-guild-log-🏡 in Pingu Support, https://discord.gg/gbxRV4Ekvh`)

    if (err) {
        var errorLink = (await PinguLibrary.errorLog(client, `pGuild Error: "${message}"`, null, err)).url;
        return pinguGuildLog.send(`[**Failed**] [**${script}**]: ${message}\n${err.message}\n\n${errorLink}\n\n<@&756383446871310399>`);
    }
    return pinguGuildLog.send(`[**Success**] [**${script}**]: ${message}`);
}
export async function pUserLog(client: Client, script: string, message: string, err ?: Error) {
    var pinguUserLog = PinguLibrary.getChannel(client, PinguLibrary.SavedServers.PinguSupport(client).id, "pingu-user-log-🧍");
    if (!pinguUserLog) return PinguLibrary.DanhoDM(client, `Couldn't get #pingu-user-log-🧍 in Pingu Support, https://discord.gg/gbxRV4Ekvh`)

    if (err) {
        var errorLink = (await PinguLibrary.errorLog(client, `pUser Error (**${script}**): "${message}"`, null, err)).url;
        return pinguUserLog.send(`[**Failed**] [**${script}**]: ${message}\n${err.message}\n\n${errorLink}\n\n<@&756383446871310399>`);
    }
    return pinguUserLog.send(`[**Success**] [**${script}**]: ${message}`);
}
export async function consoleLog(client: Client, message: string) {
    let timeFormat = `[${new Date(Date.now()).toLocaleTimeString()}]`;
    console.log(`${timeFormat} ${message}`);

    let consoleLogChannel = PinguLibrary.getChannel(client, PinguLibrary.SavedServers.PinguSupport(client).id, "console-log-📝");
    if (!consoleLogChannel) return PinguLibrary.DanhoDM(client, 'Unable to find #console-log-📝 in Pingu Support, https://discord.gg/gbxRV4Ekvh');

    consoleLogChannel.send(message);
}
export async function eventLog(client: Client, content: MessageEmbed) {
    if (client.user.id == PinguLibrary.Clients.BetaID) return;

    let eventLogChannel = PinguLibrary.getChannel(client, PinguLibrary.SavedServers.PinguSupport(client).id, "event-log-📹");
    if (!eventLogChannel) return PinguLibrary.DanhoDM(client, `Couldn't get #event-log-📹 channel in Pingu Support, https://discord.gg/gbxRV4Ekvh`)

    if (!PinguEvents.LoggedCache) PinguEvents.LoggedCache = new Array<MessageEmbed>();
    let lastCache = PinguEvents.LoggedCache[0];
    if (lastCache && (
        lastCache.description && lastCache.description == content.description ||
        lastCache.fields[0] && content.fields[0] && lastCache.fields[0].value == content.fields[0].value)
    ) return;

    PinguEvents.LoggedCache.unshift(content);
    return await eventLogChannel.send(content);
}
export async function tellLog(client: Client, sender: User, reciever: User, message: Message | MessageEmbed) {
    if (client.user.id == PinguLibrary.Clients.BetaID) return;

    var tellLogChannel = PinguLibrary.getChannel(client, PinguLibrary.SavedServers.PinguSupport(client).id, 'tell-log-💬');
    if (!tellLogChannel) return PinguLibrary.DanhoDM(client, `Couldn't get #tell-log-💬 channel in Pingu Support, https://discord.gg/gbxRV4Ekvh`)

    if ((message as object).constructor.name == "Message") {
        var messageAsMessage = message as Message;
        var consoleLog =
            messageAsMessage.content ?
                `${sender.username} sent a message to ${reciever.username} saying ` :
                messageAsMessage.attachments.array().length == 1 ?
                    `${sender.username} sent a file to ${reciever.username}` :
                    messageAsMessage.attachments.array().length > 1 ?
                        `${sender.username} sent ${messageAsMessage.attachments.array().length} files to ${reciever.username}` :
                        `${sender.username} sent something unknown to ${reciever.username}!`;

        if (messageAsMessage.content) consoleLog += messageAsMessage.content;
        if (messageAsMessage.attachments) consoleLog += messageAsMessage.attachments.map(a => `\n${a.url}`);

        PinguLibrary.consoleLog(client, consoleLog);

        var format = (ping: boolean) => `${new Date(Date.now()).toLocaleTimeString()} [<@${(ping ? sender : sender.username)}> ➡️ <@${(ping ? reciever : reciever.username)}>]`;

        if (messageAsMessage.content && messageAsMessage.attachments)
            tellLogChannel.send(format(false) + `: ||${messageAsMessage.content}||`, messageAsMessage.attachments.array())
                .then(sent => sent.edit(format(true) + `: ||${messageAsMessage.content}||`));

        else if (messageAsMessage.content)
            tellLogChannel.send(format(false) + `: ||${messageAsMessage.content}||`)
                .then(sent => sent.edit(format(true) + `: ||${messageAsMessage.content}||`));

        else if (messageAsMessage.attachments)
            tellLogChannel.send(format(false), messageAsMessage.attachments.array())
                .then(sent => sent.edit(format(true)));

        else PinguLibrary.errorLog(client, `${sender} ➡️ ${reciever} sent something that didn't have content or attachments`)
            .then(() => tellLogChannel.send(`Ran else statement - reported to ${tellLogChannel.guild.channels.cache.find(c => c.name == 'error-log')}`));
    }
    else if ((message as MessageEmbed).constructor.name == "MessageEmbed") {
        PinguLibrary.consoleLog(client, `The link between ${sender.username} & ${reciever.username} was unset.`);
        tellLogChannel.send(message as MessageEmbed)
    }
}
export async function latencyCheck(message: Message) {
    //Get latency
    let pingChannel = PinguLibrary.getChannel(message.client, PinguLibrary.SavedServers.PinguSupport(message.client).id, "ping-log-🏓");
    if (!pingChannel) return PinguLibrary.DanhoDM(message.client, `Couldn't get #ping-log-🏓 channel in Pingu Support, https://discord.gg/gbxRV4Ekvh`)

    if (message.channel == pingChannel || message.author.bot) return null;

    let pingChannelSent = await pingChannel.send(`Calculating ping`);

    let latency = pingChannelSent.createdTimestamp - message.createdTimestamp;
    pingChannelSent.edit(latency + 'ms');

    //Get outages channel
    let outages = PinguLibrary.getChannel(message.client, PinguLibrary.SavedServers.PinguSupport(message.client).id, "outages-😵");
    if (!outages) return PinguLibrary.errorLog(message.client, `Unable to find #outages-😵 channel from LatencyCheck!`);

    //Set up to find last Pingu message
    let outagesMessages = outages.messages.cache.array();
    let outageMessagesCount = outagesMessages.length - 1;

    //Find Pingu message
    for (var i = outageMessagesCount - 1; i >= 0; i--) {
        if (outagesMessages[i].author != message.client.user) continue;
        var lastPinguMessage = outagesMessages[i];
    }

    if (!lastPinguMessage) return null;

    let sendMessage = !lastPinguMessage.content.includes(`I have a latency delay on`);
    if (!sendMessage) {
        let lastMessageArgs = lastPinguMessage.content.split(` `);
        let lastLatencyExclaim = lastMessageArgs[lastMessageArgs.length - 1];
        let lastLatency = parseInt(lastLatencyExclaim.substring(0, lastLatencyExclaim.length - 1));

        if (lastLatency > 1000)
            return lastPinguMessage.edit(`I have a latency delay on ${latency}!`);
    }

    if (latency > 1000) PinguLibrary.outages(message.client, `I have a latency delay on ${latency}!`);
}
export async function raspberryLog(client: Client) {
    if (client.user.id == PinguLibrary.Clients.BetaID) return;

    let raspberryLogChannel = PinguLibrary.getChannel(client, PinguLibrary.SavedServers.PinguSupport(client).id, 'raspberry-log-🍇');
    if (!raspberryLogChannel) return PinguLibrary.DanhoDM(client, `Couldn't get #raspberry-log-🍇 channel in Pingu Support, https://discord.gg/gbxRV4Ekvh`)

    return raspberryLogChannel.send(`Pulled version ${version} from Github`);
}
