// PinguPackage/src/index.ts
export { PChannel, PClient, PGuild, PGuildMember, PItem, PMarry, PQueue, PRole, PUser } from './database';

export { GiveawayConfig, PollConfig, SuggestionConfig, ThemeConfig } from './decidable/config';
export { Giveaway, Poll, Suggestion, Theme, Decidable } from './decidable/items';
export { HandleDecidables, DecidablesTypes } from './decidable/HandleDecidables';

export { PinguCommand, CommandIDs, PinguCommandParams, PinguEvent, PinguEventParams } from './discord.js-addons'

export { EmbedField, Error, DiscordPermissions, TimeLeftObject } from './helpers';

export { PinguGuildSchema } from './MongoSchemas/PinguGuild';
export { PinguUserSchema } from './MongoSchemas/PinguUser';

export { PinguUser } from './pingu/user/PinguUser';
export { Achievement, Daily, Marry } from './pingu/user/items';

export { PinguGuild } from './pingu/guild/PinguGuild';
export { Queue, Song, ReactionRole } from './pingu/guild/items'
export { PinguLibrary } from './pingu/library/PinguLibrary';
export { PinguEvents }  from './pingu/PinguEvents';

import * as configFile from './config.json';
export const config = {
    Prefix: configFile.Prefix,
    BetaPrefix: configFile.BetaPrefix,
    token: configFile.token,
    mongoPass: configFile.mongoPass,
    api_key: configFile.api_key,
    youtube_api: configFile.youtube_api,
    google_custom_search: configFile.google_custom_search,
    version: configFile.version,
    testingMode: configFile.testingMode,
    updateStats: configFile.updateStats
}
