// PinguPackage/src/index.ts
export { PChannel, PClient, PGuild, PGuildMember, PItem, PMarry, PQueue, PRole, PUser } from './database';
export { GiveawayConfig, PollConfig, SuggestionConfig } from './decidable/config';

export { Giveaway, Poll, Suggestion } from './decidable/items';

export { EmbedField, Error, DiscordPermissions, TimeLeftObject } from './helpers';

export * as PinguGuildSchema from './MongoSchemas/PinguGuild';
export * as PinguUserSchema from './MongoSchemas/PinguUser';

export { PinguUser } from './pingu/user';
export { PinguGuild } from './pingu/guild';
export { PinguLibrary } from './pingu/library';

export { PinguEvents } from './pingu/events';


const configFile = require('../../../config.json');

class Config {
    Prefix: '*' = configFile.Prefix;
    BetaPrefix: 'b*' = configFile.BetaPrefix
    token: string = configFile.token
    mongoPass: string = configFile.mongoPass
    api_key: string = configFile.api_key
    youtube_api: string = configFile.youtube_api
    google_custom_search: string = configFile.google_custom_search
    version: 103 = configFile.version
    testingMode: false = configFile.testingMode
    updateStats: false = configFile.updateStats
}
const config = new Config();
export { config };
