// PinguPackage/src/index.ts
export { PChannel, PClient, PGuild, PGuildMember, PItem, PMarry, PQueue, PRole, PUser } from './database';
export { GiveawayConfig, PollConfig, SuggestionConfig } from './decidable/config';
export { Giveaway, Poll, Suggestion, Decidable } from './decidable/items';
export { HandleDecidables, DecidablesTypes } from './decidable/HandleDecidables';

export { EmbedField, Error, DiscordPermissions, TimeLeftObject } from './helpers';

export { PinguGuildSchema } from './MongoSchemas/PinguGuild';
export { PinguUserSchema } from './MongoSchemas/PinguUser';

export { PinguUser } from './pingu/user/PinguUser';
export { Achievement, Daily, Marry } from './pingu/user/items';

export { PinguGuild } from './pingu/guild/PinguGuild';
export { Queue, Song, ReactionRole } from './pingu/guild/items'
export { PinguLibrary } from './pingu/library/PinguLibrary';
export { PinguEvents }  from './pingu/PinguEvents';

import * as configFile from './config.json';
class Config {
    Prefix = configFile.Prefix
    BetaPrefix = configFile.BetaPrefix
    token = configFile.token
    mongoPass = configFile.mongoPass
    api_key = configFile.api_key
    youtube_api = configFile.youtube_api
    google_custom_search = configFile.google_custom_search
    version = configFile.version
    testingMode = configFile.testingMode
    updateStats = configFile.updateStats
}
const config = new Config();
export { config }
