// PinguPackage/src/pingu/guildMember/PinguGuildMember.ts
import { Collection, Guild, GuildMember, Snowflake } from "discord.js";
import { PGuildMember, PGuild } from "../../database/json";

import { cache as PinguGuildCache, GetPGuild, UpdatePGuild } from "../guild/PinguGuild";

import { pGuildLog } from "../library/PinguLibrary";
import { GuildMemberAchievementConfig, GuildMemberAchievementNotificationType } from "../achievements/config/GuildMemberAchievementConfig";

export async function WritePGuildMember(member: GuildMember, log: boolean) {
    if (member.user.bot) return null;
    
    let pGuild = await GetPGuild(member.guild);
    let pGuildMember = new PinguGuildMember(member, pGuild.settings.config.achievements.notificationTypes.members);
    pGuild.members.set(member.id, pGuildMember);
    
    if (log) pGuildLog(member.client, "guildMemberAdd", `**${member.displayName}** was successfully added to **${pGuild.name}**'s PinguGuild's members.`)
    return pGuildMember;
}
export async function GetPGuildMember(member: GuildMember) {
    let pGuild = await GetPGuild(member.guild);
    let pgm = pGuild.members.get(member.id); //Returns Mongoose.Document<PinguGuildMember>
    if (pgm) return (pgm as any).toObject ? (pgm as any).toObject() as PinguGuildMember : pgm;

    pgm = new PinguGuildMember(member, pGuild.settings.config.achievements.notificationTypes.members);
    pGuild.members.set(pgm._id, pgm);
    await UpdatePGuild(member.client, ['members'], pGuild, 'GetPGuildMember()', `Added **${pgm.name}** as a member of **${pGuild.name}**.`);
    return pgm;
}
export async function UpdatePGuildMember(member: GuildMember, pGuildMember: PinguGuildMember, scriptName: string, reason: string) {
    let pGuild = await GetPGuild(member.guild);
    pGuild.members.set(pGuildMember._id, pGuildMember);

    UpdatePGuild(member.client, ['members'], pGuild, scriptName, reason);
    return pGuildMember;
}
export async function DeletePGuildMember(member: GuildMember) {
    let pGuild = await GetPGuild(member.guild);
    pGuild.members.delete(member.id);
    UpdatePGuild(member.client, ['members'], pGuild, 'guildMemberRemove', `**${pGuild.name}**'s members, after **${member.user.tag}** left the guild.`);
    return;
}
export async function GetPGuildMembers(guild: Guild) {
    let pGuild = await GetPGuild(guild);
    return (pGuild.members as Collection<Snowflake, PinguGuildMember>).array();
}

export class PinguGuildMember extends PGuildMember {
    //#region Statics
    /**Writes and adds Member as PinguGuildMember to database and returns the new object
     * @param member Member to add as PinguGuildMember
     * @param log Whether to log it in #pGuildLog or not
     * @returns The new PinguGuildMember*/
    public static async Write(member: GuildMember, log: boolean) { return WritePGuildMember(member, log); }
    /**Get the PinguGuildMember from provided member
     * @param member Member to get as PinguGuildMember
     * @returns The provided member as PinguGuildMember*/
    public static async Get(member: GuildMember) { return GetPGuildMember(member); }
    /**Updates provided PinguGuildMember to the database
     * @param member Member to update
     * @param pGuildMember PinguGuildMember update
     * @param scriptName Script that called the method
     * @param reason Reason for updating the PinguGuildMember
     * @returns The updated PinguGuildMember*/
    public static async Update(member: GuildMember, pGuildMember: PinguGuildMember, scriptName: string, reason: string)  { return UpdatePGuildMember(member, pGuildMember, scriptName, reason); }
    /**Deletes provided member from their guild's PinguGuild's members map
     * @param member Member to delete
     * @returns what should it return?*/
    public static async Delete(member: GuildMember) { return DeletePGuildMember(member); }
    /**Get all PinguGuildMembers from specified guild
     * @param guild Guild to get all PinguGuildMembers from
     * @returns All PinguGuildMembers from specified guild*/
    public static async GetGuildMembers(guild: Guild) { return GetPGuildMembers(guild); }
    //#endregion

    constructor(member: GuildMember, achievementNotificationType: GuildMemberAchievementNotificationType) {
        super(member);
        this.guild = new PGuild(member.guild);
        this.achievementConfig = new GuildMemberAchievementConfig(achievementNotificationType);
    }

    public guild: PGuild;
    public achievementConfig: GuildMemberAchievementConfig
    //public warnings: ??
}
