// PinguPackage/src/pingu/guild/items/ReactionRole.ts
import { Message, Role, GuildChannel, MessageReaction, Client, User } from "discord.js";
import { GetPGuild } from "../PinguGuild";
import { PRole, PChannel } from "../../../database/json";
import { PermissionCheck, PermissionGranted } from "../../library/PinguLibrary";

export async function GetReactionRole(client: Client, reaction: MessageReaction, user: User) {
    let guild = reaction.message.guild;
    let pGuild = await GetPGuild(guild);
    var rr = pGuild.settings.reactionRoles.find(rr =>
        rr.messageID == reaction.message.id &&
        (rr.emoteName == reaction.emoji.name) &&
        rr.channel._id == reaction.message.channel.id
    );
    if (!rr) return null;

    let { pRole } = rr;
    let member = guild.member(user);

    let permCheck = PermissionCheck({
        author: client.user,
        client,
        channel: reaction.message.channel as GuildChannel,
        content: "No content provided"
    }, ['MANAGE_ROLES']);
    if (permCheck != PermissionGranted) {
        guild.owner.send(`I tried to give ${member.displayName} the ${pRole.name}, as ${permCheck}`);
        user.send(`I'm unable to give you the reactionrole at the moment! I've contacted ${user.username} about this.`);
        return null;
    }

    return guild.roles.fetch(pRole._id);
}
export class ReactionRole {
    constructor(message: Message, reactionName: string, role: Role) {
        this.emoteName = reactionName;
        this.pRole = new PRole(role);
        this.channel = new PChannel(message.channel as GuildChannel);
        this.messageID = message.id;
    }

    public channel: PChannel
    public messageID: string
    public emoteName: string
    public pRole: PRole

    public static async GetReactionRole(client: Client, reaction: MessageReaction, user: User) { return GetReactionRole(client, reaction, user); }
}
