// PinguPackage/src/pingu/achievements/items/AchievementBase.ts
import { PItem } from "../../../database/json/PItem";
import { PinguClientEvents } from "../../handlers/PinguEvent";

export type noGuildOnlyCommands = 
    'help' | 'info' | 'ping' | 'sort' | 'spinthewheel' | //Utility
    'fact' | 'gif' | 'meme' | 'noot' | 'daily' | 'tell' |  //Fun
    'contact' | 'invite'; //Supporting
export type guildOnlyCommands = 
    'giveaway' | 'poll' | 'suggestion' | 'serverinfo' | 'whois' | 'clear' | 'role' | 'embed' | 'fetch' | 'prefix' | 'publish' | 'reactionroles' | 'slowmode' |
    'boomer' | 'activity' | 'noice' | 'music' | 'marry' | 'quote' | 'viberate';
export type Commands = noGuildOnlyCommands | guildOnlyCommands
export type AchieverTypes = 'USER' | 'GUILD' | 'GUILDMEMBER';
export interface AchievementBaseType {
    CHAT: string,
    EVENT: keyof PinguClientEvents
    COMMAND: Commands
}

export abstract class AchievementBase extends PItem {
    constructor(id: number, name: string, description: string) {
        super({ id: id.toString(), name });
        this.description = description;
    }

    public description: string
}
