// PinguPackage/src/pingu/achievements/items/UserAchievement.ts
import { AchievementBase, AchievementBaseType } from "./AchievementBase";
import { GetPUsers } from "../../user/PinguUser";
import { Percentage } from "../../../helpers";

type noGuildOnly = 
    'help' | 'info' | 'ping' | 'sort' | 'spinthewheel' | //Utility
    'fact' | 'gif' | 'meme' | 'noot' | 'daily' | 'tell' |  //Fun
    'contact' | 'invite'; //Supporting

export interface UserAchievementType extends AchievementBaseType {
    COMMAND: noGuildOnly
}
export type UserAchievementTypeKey = keyof UserAchievementType;

export class UserAchievement<Key extends UserAchievementTypeKey> extends AchievementBase {
    constructor(id: number, name: string, key: Key, type: UserAchievementType[Key], description: string) {
        super(id, name, description);
        this.key = key;
        this.type = type;
    }

    public key: Key;
    public type: UserAchievementType[Key];

    public async getPercentage() {
        let pUsers = await GetPUsers();
        let whole = pUsers.length;
        let part = pUsers.filter(pUser => pUser.achievementConfig.achievements.find(a => a._id == this._id)).length;
        return new Percentage(whole, part);
    }

    public static Achievements: UserAchievement<UserAchievementTypeKey>[] = [
        
    ];
}
