// PinguPackage/live/helpers/DiscordPermissions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const discord_js_1 = require("discord.js");
class DiscordPermissions {
    static bitOf(value) {
        return discord_js_1.Permissions.resolve(value);
    }
}
exports.default = DiscordPermissions;
DiscordPermissions['CREATE_INSTANT_INVITE'] = 'CREATE_INSTANT_INVITE';
DiscordPermissions['KICK_MEMBERS'] = 'KICK_MEMBERS';
DiscordPermissions['BAN_MEMBERS'] = 'BAN_MEMBERS';
DiscordPermissions['ADMINISTRATOR'] = 'ADMINISTRATOR';
DiscordPermissions['MANAGE_CHANNELS'] = 'MANAGE_CHANNELS';
DiscordPermissions['MANAGE_GUILD'] = 'MANAGE_GUILD';
DiscordPermissions['ADD_REACTIONS'] = 'ADD_REACTIONS';
DiscordPermissions['VIEW_AUDIT_LOG'] = 'VIEW_AUDIT_LOG';
DiscordPermissions['PRIORITY_SPEAKER'] = 'PRIORITY_SPEAKER';
DiscordPermissions['STREAM'] = 'STREAM';
DiscordPermissions['VIEW_CHANNEL'] = 'VIEW_CHANNEL';
DiscordPermissions['SEND_MESSAGES'] = 'SEND_MESSAGES';
DiscordPermissions['SEND_TTS_MESSAGES'] = 'SEND_TTS_MESSAGES';
DiscordPermissions['MANAGE_MESSAGES'] = 'MANAGE_MESSAGES';
DiscordPermissions['EMBED_LINKS'] = 'EMBED_LINKS';
DiscordPermissions['ATTACH_FILES'] = 'ATTACH_FILES';
DiscordPermissions['READ_MESSAGE_HISTORY'] = 'READ_MESSAGE_HISTORY';
DiscordPermissions['MENTION_EVERYONE'] = 'MENTION_EVERYONE';
DiscordPermissions['USE_EXTERNAL_EMOJIS'] = 'USE_EXTERNAL_EMOJIS';
DiscordPermissions['VIEW_GUILD_INSIGHTS'] = 'VIEW_GUILD_INSIGHTS';
DiscordPermissions['CONNECT'] = 'CONNECT';
DiscordPermissions['SPEAK'] = 'SPEAK';
DiscordPermissions['MUTE_MEMBERS'] = 'MUTE_MEMBERS';
DiscordPermissions['DEAFEN_MEMBERS'] = 'DEAFEN_MEMBERS';
DiscordPermissions['MOVE_MEMBERS'] = 'MOVE_MEMBERS';
DiscordPermissions['USE_VAD'] = 'USE_VAD';
DiscordPermissions['CHANGE_NICKNAME'] = 'CHANGE_NICKNAME';
DiscordPermissions['MANAGE_NICKNAMES'] = 'MANAGE_NICKNAMES';
DiscordPermissions['MANAGE_ROLES'] = 'MANAGE_ROLES';
DiscordPermissions['MANAGE_WEBHOOKS'] = 'MANAGE_WEBHOOKS';
DiscordPermissions['MANAGE_EMOJIS'] = 'MANAGE_EMOJIS';
