// PinguPackage/live/helpers/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimeLeftObject = exports.DiscordPermissions = exports.Error = exports.BlankEmbedField = exports.EmbedField = exports.BitPermission = void 0;
var BitPermission_1 = require("./BitPermission");
Object.defineProperty(exports, "BitPermission", { enumerable: true, get: function () { return BitPermission_1.BitPermission; } });
var EmbedField_1 = require("./EmbedField");
Object.defineProperty(exports, "EmbedField", { enumerable: true, get: function () { return EmbedField_1.EmbedField; } });
Object.defineProperty(exports, "BlankEmbedField", { enumerable: true, get: function () { return EmbedField_1.BlankEmbedField; } });
var Error_1 = require("./Error");
Object.defineProperty(exports, "Error", { enumerable: true, get: function () { return Error_1.PinguError; } });
var DiscordPermissions_1 = require("./DiscordPermissions");
Object.defineProperty(exports, "DiscordPermissions", { enumerable: true, get: function () { return DiscordPermissions_1.DiscordPermissions; } });
var TimeLeftObject_1 = require("./TimeLeftObject");
Object.defineProperty(exports, "TimeLeftObject", { enumerable: true, get: function () { return TimeLeftObject_1.TimeLeftObject; } });
