// PinguPackage/live/helpers/PinguCollection.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PinguCollection = void 0;
const discord_js_1 = require("discord.js");
const Reason_1 = require("../helpers/Reason");
const Error_1 = require("../helpers/Error");
class PinguCollection {
    constructor(client, logChannelName, schemaName, newPT, typeManager) {
        this._client = client;
        this._logChannelName = logChannelName;
        this._model = require(`../MongoSchemas/${schemaName}`);
        this._schemaName = schemaName;
        this._newPT = newPT;
        this._typeManager = typeManager;
    }
    get logChannel() {
        const guild = this._client.guilds.cache.get(this._client.savedServers.get('Pingu Support').id);
        const logChannelCategory = guild.channels.cache.find(c => c.type == 'category' && c.name.includes('Pingu Logs'));
        return logChannelCategory.children.find(c => c.name.includes(this._logChannelName));
    }
    add(item, scriptName, reason) {
        return __awaiter(this, void 0, void 0, function* () {
            return this._client.DBExecute((mongoose) => __awaiter(this, void 0, void 0, function* () {
                const pItem = this._newPT(item, this._client);
                const created = yield new this._model(pItem).save();
                const _reason = new Reason_1.default('create', this._schemaName, pItem.name || pItem.tag, reason);
                this.log(scriptName, _reason, created ? null : new Error_1.default(`Unable to create ${this._schemaName}`));
                this._inner.set(item.id, pItem);
                return pItem;
            }));
        });
    }
    get(item) {
        return this._inner.get(item.id);
    }
    array() {
        return this._inner.array();
    }
    update(pItem, scriptName, reason) {
        return __awaiter(this, void 0, void 0, function* () {
            const item = yield this._typeManager(this._client).fetch(pItem._id);
            if (!item)
                throw new Error_1.default(`${this._schemaName} not found!`);
            yield this._model.updateOne({ _id: pItem._id }, pItem, null, (err) => __awaiter(this, void 0, void 0, function* () {
                const _reason = new Reason_1.default('update', this._schemaName, pItem.name || pItem.tag, reason);
                if (err)
                    return this.log(scriptName, _reason, err);
                this.log(scriptName, _reason);
                this._inner.set(pItem._id, pItem);
            }));
            return pItem;
        });
    }
    delete(item, scriptName, reason) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this._model.deleteOne({ _id: item.id }, null, err => {
                const _reason = new Reason_1.default('delete', this._schemaName, item.name || item.tag, reason);
                if (err)
                    return this.log(scriptName, _reason, err);
                this.log(scriptName, _reason);
                this._inner.delete(item.id);
            });
            return this;
        });
    }
    find(predicate) {
        const entries = this._inner.array();
        for (let i = 0; i < entries.length; i++) {
            const entry = entries[i];
            const item = this._typeManager(this._client).cache.get(entry._id);
            if (predicate(entry, item, i, this))
                return entry;
        }
        return null;
    }
    has(item) {
        return this._inner.has(item.id);
    }
    refresh(client) {
        return __awaiter(this, void 0, void 0, function* () {
            if (client)
                this._client = client;
            const dbEntries = (yield this._model.find({}).exec()).map(collDoc => collDoc.toObject());
            this._inner = new discord_js_1.Collection(dbEntries.map(entry => [entry._id, entry]));
            this._client.log('console', `Successfully refreshed entries for **${this._schemaName}**.`);
            return this;
        });
    }
    log(script, reason, err) {
        return __awaiter(this, void 0, void 0, function* () {
            const logType = this._schemaName == 'PinguGuild' ? 'pGuild' : 'pUser';
            return this._client.log(logType, script, err ? reason.errMsg : reason.succMsg, err);
        });
    }
}
exports.PinguCollection = PinguCollection;
exports.default = PinguCollection;
