// PinguPackage/live/pingu/guild/PinguGuildSettings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PinguGuildSettings = void 0;
const json_1 = require("../../database/json");
const config_1 = require("../../decidable/config");
const PinguLibrary_1 = require("../library/PinguLibrary");
class PinguGuildSettings {
    constructor(guild) {
        let welcomeChannel = guild.channels.cache.find(c => c.isText() && c.name.includes('welcome')) ||
            guild.channels.cache.find(c => c.isText() && c.name == 'general');
        if (welcomeChannel)
            this.welcomeChannel = new json_1.PChannel(welcomeChannel);
        this.giveawayConfig = new config_1.GiveawayConfig();
        this.pollConfig = new config_1.PollConfig();
        this.suggestionConfig = new config_1.SuggestionConfig();
        this.themeConfig = guild.id == PinguLibrary_1.SavedServers.DeadlyNinja(guild.client).id ? new config_1.ThemeConfig() : undefined;
        this.reactionRoles = new Array();
    }
}
exports.PinguGuildSettings = PinguGuildSettings;
