// PinguPackage/live/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.config = exports.PinguEvents = exports.PinguLibrary = exports.ReactionRole = exports.Song = exports.Queue = exports.PinguGuild = exports.Marry = exports.Daily = exports.Achievement = exports.PinguUser = exports.PinguUserSchema = exports.PinguGuildSchema = exports.TimeLeftObject = exports.DiscordPermissions = exports.Error = exports.EmbedField = exports.DecidablesTypes = exports.HandleDecidables = exports.Decidable = exports.Suggestion = exports.Poll = exports.Giveaway = exports.SuggestionConfig = exports.PollConfig = exports.GiveawayConfig = exports.PUser = exports.PRole = exports.PQueue = exports.PMarry = exports.PItem = exports.PGuildMember = exports.PGuild = exports.PClient = exports.PChannel = void 0;
var database_1 = require("./database");
Object.defineProperty(exports, "PChannel", { enumerable: true, get: function () { return database_1.PChannel; } });
Object.defineProperty(exports, "PClient", { enumerable: true, get: function () { return database_1.PClient; } });
Object.defineProperty(exports, "PGuild", { enumerable: true, get: function () { return database_1.PGuild; } });
Object.defineProperty(exports, "PGuildMember", { enumerable: true, get: function () { return database_1.PGuildMember; } });
Object.defineProperty(exports, "PItem", { enumerable: true, get: function () { return database_1.PItem; } });
Object.defineProperty(exports, "PMarry", { enumerable: true, get: function () { return database_1.PMarry; } });
Object.defineProperty(exports, "PQueue", { enumerable: true, get: function () { return database_1.PQueue; } });
Object.defineProperty(exports, "PRole", { enumerable: true, get: function () { return database_1.PRole; } });
Object.defineProperty(exports, "PUser", { enumerable: true, get: function () { return database_1.PUser; } });
var config_1 = require("./decidable/config");
Object.defineProperty(exports, "GiveawayConfig", { enumerable: true, get: function () { return config_1.GiveawayConfig; } });
Object.defineProperty(exports, "PollConfig", { enumerable: true, get: function () { return config_1.PollConfig; } });
Object.defineProperty(exports, "SuggestionConfig", { enumerable: true, get: function () { return config_1.SuggestionConfig; } });
var items_1 = require("./decidable/items");
Object.defineProperty(exports, "Giveaway", { enumerable: true, get: function () { return items_1.Giveaway; } });
Object.defineProperty(exports, "Poll", { enumerable: true, get: function () { return items_1.Poll; } });
Object.defineProperty(exports, "Suggestion", { enumerable: true, get: function () { return items_1.Suggestion; } });
Object.defineProperty(exports, "Decidable", { enumerable: true, get: function () { return items_1.Decidable; } });
var HandleDecidables_1 = require("./decidable/HandleDecidables");
Object.defineProperty(exports, "HandleDecidables", { enumerable: true, get: function () { return HandleDecidables_1.HandleDecidables; } });
Object.defineProperty(exports, "DecidablesTypes", { enumerable: true, get: function () { return HandleDecidables_1.DecidablesTypes; } });
var helpers_1 = require("./helpers");
Object.defineProperty(exports, "EmbedField", { enumerable: true, get: function () { return helpers_1.EmbedField; } });
Object.defineProperty(exports, "Error", { enumerable: true, get: function () { return helpers_1.Error; } });
Object.defineProperty(exports, "DiscordPermissions", { enumerable: true, get: function () { return helpers_1.DiscordPermissions; } });
Object.defineProperty(exports, "TimeLeftObject", { enumerable: true, get: function () { return helpers_1.TimeLeftObject; } });
var PinguGuild_1 = require("./MongoSchemas/PinguGuild");
Object.defineProperty(exports, "PinguGuildSchema", { enumerable: true, get: function () { return PinguGuild_1.PinguGuildSchema; } });
var PinguUser_1 = require("./MongoSchemas/PinguUser");
Object.defineProperty(exports, "PinguUserSchema", { enumerable: true, get: function () { return PinguUser_1.PinguUserSchema; } });
var PinguUser_2 = require("./pingu/user/PinguUser");
Object.defineProperty(exports, "PinguUser", { enumerable: true, get: function () { return PinguUser_2.PinguUser; } });
var items_2 = require("./pingu/user/items");
Object.defineProperty(exports, "Achievement", { enumerable: true, get: function () { return items_2.Achievement; } });
Object.defineProperty(exports, "Daily", { enumerable: true, get: function () { return items_2.Daily; } });
Object.defineProperty(exports, "Marry", { enumerable: true, get: function () { return items_2.Marry; } });
var PinguGuild_2 = require("./pingu/guild/PinguGuild");
Object.defineProperty(exports, "PinguGuild", { enumerable: true, get: function () { return PinguGuild_2.PinguGuild; } });
var items_3 = require("./pingu/guild/items");
Object.defineProperty(exports, "Queue", { enumerable: true, get: function () { return items_3.Queue; } });
Object.defineProperty(exports, "Song", { enumerable: true, get: function () { return items_3.Song; } });
Object.defineProperty(exports, "ReactionRole", { enumerable: true, get: function () { return items_3.ReactionRole; } });
var PinguLibrary_1 = require("./pingu/library/PinguLibrary");
Object.defineProperty(exports, "PinguLibrary", { enumerable: true, get: function () { return PinguLibrary_1.PinguLibrary; } });
var PinguEvents_1 = require("./pingu/PinguEvents");
Object.defineProperty(exports, "PinguEvents", { enumerable: true, get: function () { return PinguEvents_1.PinguEvents; } });
const configFile = require("./config.json");
class Config {
    constructor() {
        this.Prefix = configFile.Prefix;
        this.BetaPrefix = configFile.BetaPrefix;
        this.token = configFile.token;
        this.mongoPass = configFile.mongoPass;
        this.api_key = configFile.api_key;
        this.youtube_api = configFile.youtube_api;
        this.google_custom_search = configFile.google_custom_search;
        this.version = configFile.version;
        this.testingMode = configFile.testingMode;
        this.updateStats = configFile.updateStats;
    }
}
const config = new Config();
exports.config = config;
