// PinguPackage/src/pingu/guild/PinguGuildSettings.ts
import { Guild } from 'discord.js';

import { PChannel } from '../../database/json';
import { GiveawayConfig, PollConfig, SuggestionConfig, ThemeConfig } from '../../decidable/config'
import { SavedServers } from '../library/PinguLibrary'

import { ReactionRole } from './items';

export class PinguGuildSettings {
    constructor(guild: Guild) {
        let welcomeChannel = guild.channels.cache.find(c => c.isText() && c.name.includes('welcome')) ||
            guild.channels.cache.find(c => c.isText() && c.name == 'general');

        if (welcomeChannel) this.welcomeChannel = new PChannel(welcomeChannel);

        this.giveawayConfig = new GiveawayConfig();
        this.pollConfig = new PollConfig();
        this.suggestionConfig = new SuggestionConfig();
        this.themeConfig = guild.id == SavedServers.DeadlyNinja(guild.client).id ? new ThemeConfig() : undefined;

        this.reactionRoles = new Array<ReactionRole>();
    }

    public welcomeChannel: PChannel;

    public giveawayConfig: GiveawayConfig;
    public pollConfig: PollConfig;
    public suggestionConfig: SuggestionConfig;
    public themeConfig: ThemeConfig;

    public reactionRoles: ReactionRole[];
}
