// PinguPackage/live/pingu/guild/items/music/Song.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const database_1 = require("../../../../database");
const helpers_1 = require("../../../../helpers");
class Song {
    constructor(author, songInfo) {
        //YouTube
        this.link = songInfo.video_url;
        this.title = songInfo.title;
        this.author = songInfo.author && songInfo.author.name;
        this.length = this.GetLength(songInfo.lengthSeconds);
        this.lengthMS = parseInt(songInfo.lengthSeconds) * 1000;
        this.thumbnail = songInfo.thumbnails[0].url;
        this.requestedBy = new database_1.PUser(author);
        this._id = 0;
        this.volume = -1;
        this.loop = false;
        this.endsAt = null;
    }
    play() {
        this.endsAt = new Date(Date.now() + this.lengthMS);
    }
    stop() {
        this.endsAt = null;
    }
    getTimeLeft() {
        return new helpers_1.TimeLeftObject(new Date(Date.now()), this.endsAt);
    }
    GetLength(secondsLength) {
        var seconds = parseInt(secondsLength), minutes = 0, hours = 0, final = [];
        if (seconds > 59) {
            while (seconds > 59) {
                seconds -= 60;
                minutes++;
            }
        }
        if (minutes > 59) {
            while (minutes > 59) {
                minutes -= 60;
                hours++;
            }
        }
        final.push(hours, minutes, seconds);
        return final.map(i => i < 10 ? `0${i}` : i).join('.');
    }
}
exports.default = Song;
