// PinguPackage/src/pingu/user/PinguUser.ts
import { Client, User } from "discord.js";
import { PUser, PQueue, PGuild } from "../../database";
import { Marry, Daily, Achievement } from "./items";

import { PinguUserSchema } from '../../MongoSchemas/PinguUser';
import { DBExecute, pUserLog } from '../library/PinguLibrary'

import { Error } from "../../helpers";

export async function WritePUser(client: Client, user: User, scriptName: string, succMsg: string, errMsg: string) {
    DBExecute(client, async mongoose => {
        let created = await new PinguUserSchema(new PinguUser(user)).save();
        if (!created) pUserLog(client, scriptName, errMsg);
        else pUserLog(client, scriptName, succMsg);
    });
}
export async function GetPUser(user: User): Promise<PinguUser> {
    let pUserDoc = await PinguUserSchema.findOne({ _id: user.id }).exec();
    return pUserDoc ? pUserDoc.toObject() : null;
}
export async function UpdatePUser(client: Client, updatedProperty: object, pUser: PinguUser, scriptName: string, succMsg: string, errMsg: string): Promise<PinguUser> {
    return await PinguUserSchema.updateOne({ _id: pUser._id }, updatedProperty, null, err => {
        if (err) pUserLog(client, scriptName, errMsg, err);
        else pUserLog(client, scriptName, succMsg);
    })
}
export async function DeletePUser(client: Client, user: User, scriptName: string, succMsg: string, errMsg: string): Promise<PinguUser> {
    return await PinguUserSchema.deleteOne({ _id: user.id }, null, err => {
        if (err) pUserLog(client, scriptName, errMsg, new Error(err));
        else pUserLog(client, scriptName, succMsg);
    });
}

export async function GetPUsers(): Promise<PinguUser[]> {
    return (await PinguUserSchema.find({}).exec()).map(collDoc => collDoc.toObject());
}


export class PinguUser {
    public static async WritePUser(client: Client, user: User, scriptName: string, succMsg: string, errMsg: string) {
        return WritePUser(client, user, scriptName, succMsg, errMsg)
    }
    public static async GetPUser(user: User): Promise<PinguUser> {
        return GetPUser(user);
    }
    public static async UpdatePUser(client: Client, updatedProperty: object, pUser: PinguUser, scriptName: string, succMsg: string, errMsg: string): Promise<PinguUser> {
        return UpdatePUser(client, updatedProperty, pUser, scriptName, succMsg, errMsg);
    }
    public static async DeletePUser(client: Client, user: User, scriptName: string, succMsg: string, errMsg: string): Promise<PinguUser> {
        return DeletePUser(client, user, scriptName, succMsg, errMsg);
    }
    public static async GetPUsers(): Promise<PinguUser[]> {
        return GetPUsers();
    }

    constructor(user: User) {
        let pUser = new PUser(user);
        this._id = pUser._id;
        this.tag = pUser.name;
        this.sharedServers = user.client.guilds.cache.filter(g => g.members.cache.has(user.id)).map(g => new PGuild(g));
        this.marry = new Marry();
        this.replyPerson = null;
        this.daily = new Daily();
        this.avatar = user.avatarURL();
        this.playlists = new Array<PQueue>();
    }
    public _id: string
    public tag: string
    public sharedServers: PGuild[]
    public marry: Marry
    public replyPerson: PUser
    public daily: Daily
    public avatar: string
    public playlists: PQueue[]
    //public Achievements: Achievement[]
}
