// PinguPackage/src/pingu/handlers/PinguCommand.ts
import { Message, PermissionString, VoiceConnection } from 'discord.js';

import { PinguHandler } from './PinguHandler'

import { PClient } from '../../database/json/PClient';
import { DiscordPermissions } from '../../helpers'

import { errorLog } from "../library/PinguLibrary";
import { PinguUser } from '../user/PinguUser';
import { PinguGuild } from '../guild/PinguGuild';

export enum CommandCategories { 'Utility', 'Fun', 'Supporting', 'DevOnly', 'GuildSpecific' }
type CommandCategoriesType = keyof typeof CommandCategories;

export interface PinguCommandParams {
    message: Message,
    args?: string[],
    pAuthor?: PinguUser,
    pGuild?: PinguGuild,
    pGuildClient?: PClient
}

export class PinguCommand extends PinguHandler {
    constructor(name: string, category: CommandCategoriesType, description: string, data: {
        usage: string, 
        guildOnly?: false,
        specificGuildID?: string,
        examples?: string[],
        permissions: PermissionString[],
        aliases?: string[],
    }, 
    execute: (params: PinguCommandParams) => Promise<void | Message | VoiceConnection>) 
    { 
        //Must need these
        super(name);
        this.description = description;
        this.category = category;
        if (execute) this.execute = execute;

        if (data) {
            const { permissions } = data;
            this.permissions = permissions && permissions.length ? [...permissions, DiscordPermissions.SEND_MESSAGES] : [DiscordPermissions.SEND_MESSAGES];
            
            //Optional
            const { usage, guildOnly, specificGuildID, examples, aliases } = data;
            this.usage = usage || "";
            this.guildOnly = guildOnly;
            this.specificGuildID = specificGuildID;
            this.examples = examples && examples.length ? examples : [""];
            this.aliases = aliases && aliases.length ? aliases : undefined;
        }
    }
    
    public description: string;
    public usage: string;
    public guildOnly = false;
    public category: CommandCategoriesType;
    public specificGuildID: string;
    public examples: string[];
    public permissions: string[];
    public aliases: string[];
    
    public async execute(params: PinguCommandParams): Promise<void | Message | VoiceConnection> {
        return errorLog(params.message.client, `Execute for command **${this.name}**, was not defined!`);
    }
}
