// PinguPackage/live/pingu/library/developers.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isPinguDev = exports.CacheDevelopers = exports.Developers = void 0;
const discord_js_1 = require("discord.js");
exports.Developers = new discord_js_1.Collection();
function CacheDevelopers(client) {
    return __awaiter(this, void 0, void 0, function* () {
        const developers = {
            Danho: '245572699894710272',
            SynthySytro: '405331883157880846',
            Slothman: '290131910091603968',
            DefilerOfCats: '803903863706484756'
        };
        const devIDs = Object.keys(developers).map(dev => developers[dev]);
        for (const id of devIDs) {
            let user = yield client.users.fetch(id);
            exports.Developers.set(id, user);
        }
        return exports.Developers;
    });
}
exports.CacheDevelopers = CacheDevelopers;
function isPinguDev(user) {
    return exports.Developers.get(user.id) != null;
}
exports.isPinguDev = isPinguDev;
