// PinguPackage/live/pingu/library/PinguLibrary.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PinguLibrary = exports.AchievementCheck = exports.AchievementCheckType = void 0;
const UserAchievement_1 = require("../achievements/items/UserAchievement");
const GuildMemberAchievement_1 = require("../achievements/items/GuildMemberAchievement");
const GuildAchievement_1 = require("../achievements/items/GuildAchievement");
const PinguUser_1 = require("../user/PinguUser");
const PinguGuildMember_1 = require("../guildMember/PinguGuildMember");
const PAchievement_1 = require("../../database/json/PAchievement");
const UserAchievementConfig_1 = require("../achievements/config/UserAchievementConfig");
const GuildMemberAchievementConfig_1 = require("../achievements/config/GuildMemberAchievementConfig");
const GuildAchievementConfig_1 = require("../achievements/config/GuildAchievementConfig");
function AchievementCheckType(client, achieverType, achiever, key, keyType, config, callbackKey, callback) {
    return __awaiter(this, void 0, void 0, function* () {
        const filter = (arr) => arr.filter(i => i.key == key && i.type == keyType);
        let allAchievements = filter((function getAllAchievements() {
            switch (achieverType) {
                case 'USER': return UserAchievement_1.UserAchievement.Achievements;
                case 'GUILDMEMBER': return GuildMemberAchievement_1.GuildMemberAchievement.Achievements;
                case 'GUILD': return GuildAchievement_1.GuildAchievement.Achievements;
                default: return null;
            }
        })());
        if (!allAchievements)
            return null;
        let pAchievements = ((yield (function getAllPAchievements() {
            return __awaiter(this, void 0, void 0, function* () {
                switch (achieverType) {
                    case 'USER':
                        const user = yield PinguUser_1.GetPUser(achiever);
                        return user === null || user === void 0 ? void 0 : user.achievementConfig.achievements;
                    case 'GUILDMEMBER':
                        const member = yield PinguGuildMember_1.GetPGuildMember(achiever, 'PinguLibrary.AchievementCheckType()');
                        return member === null || member === void 0 ? void 0 : member.achievementConfig.achievements;
                    case 'GUILD': return (yield GetPGuild(achiever)).settings.config.achievements.achievements;
                    default: return null;
                }
            });
        })()) || []).map(pa => pa._id);
        //Find an achievement matching Key & Type, that achiever doesn't have, and the achievement's callback returns true
        let achievements = yield (function Find() {
            return __awaiter(this, void 0, void 0, function* () {
                const result = [];
                for (const a of allAchievements)
                    if (!pAchievements.includes(a._id) && (yield a.callback(callback)))
                        result.push(a);
                return result;
            });
        })();
        if (!achievements.length)
            return null;
        return (yield Promise.all(achievements.map((achievement) => __awaiter(this, void 0, void 0, function* () {
            let pAchievement = new PAchievement_1.default({ _id: achievement._id, achievedAt: new Date() });
            try {
                const notificationType = config.notificationType || config.notificationTypes.guild;
            }
            catch (err) {
                config = (function newConfig() {
                    switch (achieverType) {
                        case 'USER': return new UserAchievementConfig_1.default('NONE');
                        case 'GUILDMEMBER': return new GuildMemberAchievementConfig_1.default('NONE');
                        case 'GUILD': return new GuildAchievementConfig_1.default({ guild: 'NONE', members: 'NONE' }, achiever.id);
                        default: return null;
                    }
                })();
            }
            yield (function UpdateDB() {
                return __awaiter(this, void 0, void 0, function* () {
                    const scriptName = 'PinguLibrary.AchievementCheckType()';
                    switch (achieverType) {
                        case 'USER':
                            let pUser = yield PinguUser_1.GetPUser(achiever);
                            pUser.achievementConfig.achievements.push(pAchievement);
                            return PinguUser_1.UpdatePUser(client, ['achievementConfig'], pUser, scriptName, `Added ${achieverType.toLowerCase()} achievement, #${pAchievement._id} "${achievement.name}" to **${achiever.tag}**'s PinguUser achievements collection`);
                        case 'GUILDMEMBER':
                            let pGuildMember = yield PinguGuildMember_1.GetPGuildMember(achiever, scriptName);
                            pGuildMember.achievementConfig.achievements.push(pAchievement);
                            return PinguGuildMember_1.UpdatePGuildMember(achiever, pGuildMember, scriptName, `Added ${achieverType.toLowerCase()} achievement, #${pAchievement._id} "${achievement.name}" to **${achiever.user.tag}**'s PinguGuildMember achievements collection`);
                        case 'GUILD':
                            let pGuild = yield GetPGuild(achiever);
                            pGuild.settings.config.achievements.achievements.push(pAchievement);
                            return UpdatePGuild(client, ['settings'], pGuild, scriptName, `Added ${achieverType.toLowerCase()} achievement, #${pAchievement._id} "${achievement.name}" to **${achiever.name}**'s PinguGuild achievements collection`);
                        default: return null;
                    }
                });
            })();
            return (function notify() {
                switch (achieverType) {
                    case 'USER': return UserAchievementConfig_1.default.notify(client, achiever, achievement, config);
                    case 'GUILDMEMBER': return GuildMemberAchievementConfig_1.default.notify(client, achiever, achievement, config);
                    case 'GUILD': return GuildAchievementConfig_1.default.notify(client, achiever, achievement, config);
                    default: return null;
                }
            })();
        }))))[0];
    });
}
exports.AchievementCheckType = AchievementCheckType;
function AchievementCheck(client, data, key, type, callback) {
    var _a;
    return __awaiter(this, void 0, void 0, function* () {
        if (data.user && !data.user.bot) {
            let pUser = yield PinguUser_1.GetPUser(data.user);
            if (!pUser)
                return false;
            var givenAchievement = yield AchievementCheckType(client, 'USER', data.user, key, type, pUser.achievementConfig, key, callback);
        }
        if (data.guild) {
            let pGuild = yield GetPGuild(data.guild);
            if (!pGuild)
                return false;
            givenAchievement = yield AchievementCheckType(client, 'GUILD', data.guild, key, type, pGuild.settings.config.achievements, key, callback);
        }
        if (data.guildMember && !((_a = data.guildMember.user) === null || _a === void 0 ? void 0 : _a.bot)) {
            let pGuildMember = yield PinguGuildMember_1.GetPGuildMember(data.guildMember, 'PinguLibrary.AchievementCheck()');
            if (!pGuildMember)
                return false;
            givenAchievement = yield AchievementCheckType(client, 'GUILDMEMBER', data.guildMember, key, type, pGuildMember.achievementConfig, key, callback);
        }
        return givenAchievement != null;
    });
}
exports.AchievementCheck = AchievementCheck;
//#endregion
//#region Badges
const PinguBadge_1 = require("../badge/PinguBadge");
//#endregion
class PinguLibrary {
    //#region Achievement
    static AchievementCheck(client, data, key, type, callbackParams) {
        return __awaiter(this, void 0, void 0, function* () {
            return AchievementCheck(client, data, key, type, callbackParams);
        });
    }
    static SetBadges() { return PinguBadge_1.SetBadges(); }
    static getBadges(user) { return PinguBadge_1.getBadges(user); }
}
exports.PinguLibrary = PinguLibrary;
//#endregion
//#region Badges
PinguLibrary.Badges = PinguBadge_1.Badges;
exports.default = PinguLibrary;
